crates/prune/prune/src/segments/history.rs

```diff
@@ -10,103 +10,158 @@ use reth_db_api::{
+        }
+
+        while let Some((key, value)) = cursor
+            .next()?
+            .map(|(k, v)| Result::<_, DatabaseError>::Ok((k.key()?, v)))
+            .transpose()?
+        {
+            if key_matches(&key, &sharded_key) {
+                match prune_shard(&mut cursor, key, value, to_block, &key_matches)? {
+                    PruneShardOutcome::Deleted => deleted += 1,
+                    PruneShardOutcome::Updated => updated += 1,
+                    PruneShardOutcome::Unchanged => unchanged += 1,
+                }
+            } else {
+                break
             }
         }
-        // Shard contains block numbers that are higher than the target one, so we need to
-        // filter it. It is guaranteed that further shards for this sharded key will not
-        // contain the target block number, as it's in this shard.
-        else {
-            let higher_blocks =
-                blocks.iter().skip_while(|block| *block <= to_block).collect::<Vec<_>>();
+    }
 
-            // If there were blocks less than or equal to the target one
-            // (so the shard has changed), update the shard.
-            if blocks.len() as usize != higher_blocks.len() {
-                // If there will be no more blocks in the shard after pruning blocks below target
-                // block, we need to remove it, as empty shards are not allowed.
-                if higher_blocks.is_empty() {
-                    if key.as_ref().highest_block_number == u64::MAX {
-                        let prev_row = cursor.prev()?;
-                        match prev_row {
-                            // If current shard is the last shard for the sharded key that
-                            // has previous shards, replace it with the previous shard.
-                            Some((prev_key, prev_value)) if key_matches(&prev_key, &key) => {
-                                cursor.delete_current()?;
-                                deleted += 1;
-                                // Upsert will replace the last shard for this sharded key with
-                                // the previous value.
-                                cursor.upsert(key.clone(), prev_value)?;
-                            }
-                            // If there's no previous shard for this sharded key,
-                            // just delete last shard completely.
-                            _ => {
-                                // If we successfully moved the cursor to a previous row,
-                                // jump to the original last shard.
-                                if prev_row.is_some() {
-                                    cursor.next()?;
-                                }
-                                // Delete shard.
-                                cursor.delete_current()?;
-                                deleted += 1;
+    Ok((deleted, updated, unchanged))
```

Comment:
better type safety to return the `PruneShardOutcome` type, so that no bugs in future can be made mixing up the undescriptive tuple indices, e.g. mix up `deleted` and `updated` numbers

Reply:
agreed, done